src/lib/ipc/shmem/shmem.c

```diff
@@ -0,0 +1,38 @@
+#include <sys/mman.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <stdbool.h>
+
+bool shmem_unmap(void *mem, size_t size) {
+  if (munmap(mem, size) == -1) {
+    perror("munmap");
+    return(false);
+  }
+  return(true);
+}
+
+// Note: the io.* file handle passed to us is converted to a Unix
+// filehandle (FILE *) by LuaJIT
+char *shmem_grow(void *fh, void *old_mem, size_t old_size, size_t new_size) {
+  int fd;
+  void *mem;
+
+  if (old_mem != NULL) {
```

Comment:
Is this function vulnerable to creating dangling pointers due to remapping memory? If so, could we fix that?

Example:

1. Lua code calls `counter` module and gets the address of a counter for repeated use.
2. `munmap()` invalidates the old mapping and then `mmap()` returns a different address (for whatever reason).
3. Lua code tries to use the remembered counter value and causes a `SIGSEGV` because it is no longer mapped.

Is this a valid error scenario? (I do like the idea of counters having stable virtual addresses: then I expect they can be incremented very cheaply indeed and so we need not be shy of doing so in the fast path).

If indeed this would be a problem, could we solve it simply by skipping the `munmap()`? Then even if `mmap()` happened to choose a new address then all previous mappings would remain valid as separate mappings?